### PR TITLE
[Feature] Component structure and management of states and styles

### PR DESCRIPTION
**Pull request by jorGonzalez291292**
## Useful links
- [Contributing](https://github.com/mercadolibre/fury_andesui-android/blob/develop/CONTRIBUTING.md) has more information and tips for a great pull request. This will give you a nice introduction about how to contribute with Andes.
- [Wiki](https://github.com/mercadolibre/fury_andesui-android/wiki) has all the info you need to start with Andes. Even I heard there's a video there...

## Motivation
Create a new component that allows us to write a code.

## Description
Component structure is created following the architecture definitions, along with the handling of states and styles.

## Affected component
AndesTextfield is affected. An internal method is added that allows you to change the alignment of the text, and the visibility behavior of iconComponent is fixed.

## Frontify component link
The UX team behind Andes keeps all the Andes Components inside the Frontify site. Each component has its own section. Let us know the section of the component you want to modify.

## Screenshots
Screenshot_1600551797
Screenshot_1600620909
Screenshot_1600551683

## Checks
Are you sure you followed all those steps? In such case, let's say with me "I solemnly declare that ...":
- [x] I have coded an example inside the Demo App,
- [ ] I have added proper tests,
- [ ] I have modified the [Changelog](https://github.com/mercadolibre/fury_andesui-android/blob/develop/CHANGELOG.md) file,
- [ ] I have updated GitHub wiki,
- [ ] I have the explicit approval of one or more members of the UX Team,
- [ ] I have updated the version in gradle.properties file.

## Change type
This is easy, let us know what this code is about:
- [x] This code adds a new component
- [x] This code includes improvements to an existent component
- [ ] This code improves or modifies ONLY the Demo App.

## Check common errors
Whether you are an Android or iOS developer and if you're still there then we'll give you a present:
https://proandroiddev.com/how-to-maximize-androids-ui-reusability-5-common-mistakes-cb2571216a9f

### CODE REVIEW

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+     * Getter and setter for the [state].
+     */
+    var state: AndesTextfieldCodeState
+        get() = andesTextfieldCodeAttrs.state
+        set(value) {
+            andesTextfieldCodeAttrs = andesTextfieldCodeAttrs.copy(state = value)
+            val config = createConfig()
+            setupEnabledView()
+            setupColorComponents(config)
+            setupHelperComponent(config)
```
Comment (haurbano) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt]: Por qué es necesario al settear el state volver a hacer el setup del helperComponent?

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+
+    /**
+     * Getter and setter for [helper].
+     */
+    var helper: String?
+        get() = andesTextfieldCodeAttrs.helper
+        set(value) {
+            andesTextfieldCodeAttrs = andesTextfieldCodeAttrs.copy(helper = value)
+            val config = createConfig()
+            setupColorComponents(config)
```
Comment (haurbano) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt]: Por qué es necesario hacer el setup de colorComponent al settear este helper?

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/style/AndesTextfieldCodeStyle.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
+package com.mercadolibre.android.andesui.textfield.style
+
+import java.util.*
```
Comment (haurbano) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/style/AndesTextfieldCodeStyle.kt]: Necesitamos todo lo que está dentro de java.util o podemos remplazar el wildcard por lo que necesitamos solamente?

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+        textfieldBoxCodeContainer.removeAllViews()
+        val boxesIterator = config.boxesPattern.iterator()
+        while (boxesIterator.hasNext()) {
+            var boxes = boxesIterator.next()
+            while (boxes-- > 0) {
+                setUpBoxView(config)
+                setupMarginBetweenBoxes(config, boxes == 0)
+            }
+        }
+        text = currentText?.takeIf { it.isNotEmpty() }
```
Comment (cgaggino) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt]: No entiendo lo de current text en este contexto

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt (hunk 1 above)
Comment (jorGonzalez291292) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt]: Tenés razón, es innecesario! Eliminado.

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt (hunk 2 above)
Comment (jorGonzalez291292) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt]: idem al anterior.

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/style/AndesTextfieldCodeStyle.kt (hunk 3 above)
Comment (jorGonzalez291292) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/style/AndesTextfieldCodeStyle.kt]: Ya quedo optimizado. Gracias!

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt (hunk 4 above)
Comment (jorGonzalez291292) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt]: Updateado!

**Review by haurbano — APPROVED** [review]

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    private lateinit var andesTextfieldCodeAttrs: AndesTextfieldCodeAttrs
+    private lateinit var textfieldCodeContainer: ConstraintLayout
+    private lateinit var textfieldBoxCodeContainer: LinearLayout
+    private lateinit var labelComponent: TextView
+    private lateinit var helperComponent: TextView
+    private lateinit var iconComponent: SimpleDraweeView
+    private var currentText: String? = null
+
+    constructor(context: Context?) : super(context) {
+        initAttrs(LABEL_DEFAULT, HELP_LABEL_DEFAULT, STYLE_DEFAULT, STATE_DEFAULT)
```
Comment (jperezgonzal) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt]: Es necesario tener este constuctor? No se podrian poner como opcionales y con un valor por defecto en el constructor de abajo?

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+
+    private fun setupViewAsClickable() {
+        isFocusable = true
+        textfieldBoxCodeContainer.isClickable = true
+        textfieldBoxCodeContainer.isFocusable = true
+    }
+
+    private fun setupEnabledView() {
+        if (state == AndesTextfieldCodeState.DISABLED) {
+            isEnabled = false
```
Comment (jperezgonzal) [inline components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt]: Llevaría el **isEnabled** para la configuración y sacaría el **if**.

File: components/src/main/res/textfield/values/attrs.xml [hunk 7]
```diff
@@ -57,4 +57,24 @@
...
+
+        <attr name="andesTextfieldCodeLabel" format="reference" />
+
+        <attr name="andesTextfieldCodeHelper" format="reference" />
+
+        <attr name="andesTextfieldCodeStyle">
+            <enum name="threesome" value="10" />
+            <enum name="foursome" value="11" />
+            <enum name="three_by_three" value="12" />
+        </attr>
```
Comment (jperezgonzal) [inline components/src/main/res/textfield/values/attrs.xml:71]: Estos valores lo pondria mas distantes como en los otros componentes.

Por ejemplo, para el style 1000, 1001, 1002 y para el state 2000, 2001, 2002

Comment (jorGonzalez291292) [reply components/src/main/res/textfield/values/attrs.xml:71]: Dale!

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt (hunk 6 above)
Comment (jorGonzalez291292) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt]: Excelente, cambiado!

File: components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt (hunk 5 above)
Comment (jorGonzalez291292) [reply components/src/main/java/com/mercadolibre/android/andesui/textfield/AndesTextfieldCode.kt]: Tenes razón, había replicado los constructores de `AndesTextfield` jaja

**Review by jperezgonzal — APPROVED** [review]

**Pull request merged**